src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -675,6 +810,24 @@
       jruby-request/wrap-with-environment-validation
       jruby-request/wrap-with-error-handling))
 
+(schema/defn ^:always-validate
+  all-plans-handler :- IFn
+  "Handler for processing an incoming all_plans Ring request"
+  [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
+  (-> (all-plans-fn jruby-service)
+      (jruby-request/wrap-with-jruby-instance jruby-service)
```

Comment:
So, if we get the jruby instance here:

Does this mean all processing - including data transformation outside of the ruby calls - is done with a jruby instance lock? I don't know how much this matters, but if that is the case that would mean we are getting a lock for slightly longer than we need, since all we really care about is the one call to `.GetPlans` (or `.GetPlanData` for `plan-details-handler`)

I don't know how _much_ that would effect performance: It's not a huge amount of processing that we do outside the call to Jruby. However, since we are adding code to share jruby instances with compilation, maybe every small amount of time counts?

I can see an argument that the small amount of processing done here doesn't really matter, I just want to be pedantic and careful about resource sharing with the jruby instances.

Reply:
This is how all of puppetserver does it - jruby checkout is an HTTP middleware